### PR TITLE
lua: rewrite crc32 digest via Lua C API

### PR DESCRIPTION
**Pull request by igormunkin**
There are some issues with crc32 digest builtins implemented via FFI, so
they are rewritted line by line via Lua C API.

NO_DOC=no visible changes
NO_CHANGELOG=no visible changes

### CODE REVIEW

File: src/lua/digest.c [hunk 1]
```diff
@@ -90,13 +91,174 @@ lua_pbkdf2(lua_State *L)
...
+{
+	/* Get <self.value>. */
+	lua_getfield(L, 1, "value");
+	return 1;
+}
+
+int
+crc32_methods_update(lua_State *L)
+{
+	uint32_t crc32_begin, crc32_result;
```
Comment (Buristan) [inline src/lua/digest.c:107]: In Tarantool there is no need to declare variables in the start of block. You may just declare them below.

Here and below.

**Review by Buristan — CHANGES_REQUESTED** [review]
> lua: rewrite crc32 digest via Lua C API
>
> There are some issues with crc32 digest builtins implemented via FFI, so
> they are rewritted line by line via Lua C API.

I agree with Sasha's comment about commit message -- please, make it more specific.

Also, please add test (even creepy one) to verify this patch (yes, we know that it is working from customer feedback, but whatever).

Side note: Can we just turn off JIT for suspicious functions in digest.lua, how it was done for luafun? It looks easier solution, doesn't it?

Also, I suggest to create ticket with repro and problem description (and mention this tickett here). It issue should be investigated due to possible bug in LuaJIT.

The patch itself LGTM, except 2 nits.

File: src/lua/digest.c [hunk 2]
```diff
@@ -90,13 +91,174 @@ lua_pbkdf2(lua_State *L)
...
+	return 0;
+}
+
+int
+crc32_methods_clear(lua_State *L)
+{
+	/* Get <CRC32> upvalue. */
+	lua_pushvalue(L, lua_upvalueindex(1));
+	/* Get <CRC32.crc_begin>. */
+	lua_getfield(L, -1, "crc_begin");
```
[the remark this answers is not in the document]
Comment (igormunkin) [reply src/lua/digest.c]: Then internal structure will be changed. Anyway, this part is left intact in the [new version](https://github.com/tarantool/tarantool/commit/b2c3f868e98f5ff35469bf6a7d06a6f25b682f8e).

**Comment by igormunkin** [comment]
@Buristan, thanks for your review!

> I agree with Sasha's comment about commit message -- please, make it more specific.

Fixed, see the new version below:
```
lua: rewrite crc32 digest via Lua C API

As a result of recording <crc32:update> method or <digest.crc32>
function wrong semantics is compiled (strictly saying, the resulting
trace produces the different result from the one yielded by
interpreter). The easiest solution is disabling JIT for particular
functions, however, such approach drops the overall platform
performance. Hence, the mentioned functions are rewritten line by line
via Lua C API to avoid JIT misbehaviour.
```

> Also, please add test (even creepy one) to verify this patch (yes, we know that it is working from customer feedback, but whatever).

Added the one, you've reduced some time before.

> Side note: Can we just turn off JIT for suspicious functions in digest.lua, how it was done for luafun? It looks easier solution, doesn't it?

I've re-implemented the patch this way. Anyway, we didn't disable JIT engine for LuaFun functions in Tarantool, but only for a single particular flaky test. Furthermore, such solution can lead to so-called blacklist cancer, that looks less
 better solution than stitching. For now I have no specific proofs and benchmarks, but here are two poor man ones you can find below.
 
**TL;DR**: jit.off drops the overall performance, so Lua C approach is chosen.
<details>

* jit.off implementation ([igormunkin:97205dd07](https://github.com/tarantool/tarantool/commit/97205dd07ca9f3ea6cbf52c5438f6c589d4ff6bc))
* Lua C implementation ([igormunkin:b2c3f868e](https://github.com/tarantool/tarantool/commit/b2c3f868e98f5ff35469bf6a7d06a6f25b682f8e))

```
$ cat test.lua
local clock = require('clock')
local digest = require('digest')

local function strcrc32(shard_key)
  if type(shard_key) ~= 'table' then
    return digest.crc32(tostring(shard_key))
  else
    local crc32 = digest.crc32.new()
    for _, v in ipairs(shard_key) do
      crc32:update(tostring(v))
    end
    return crc32:result()
  end
end

local t = { }

local start = clock.monotonic()
for i = 1, 1e6 do
  t[i] = strcrc32({'db095e64-9972-400b-a65b-d44047fcb812', nil})
end
local finish = clock.monotonic() - start
print(finish)
$ ./src/tarantool -v && for i in $(seq 1 10); do ./src/tarantool test.lua; done # jit.off implementation
Tarantool 2.10.0-beta2-207-g97205dd07
Target: Linux-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=ON
Compiler: /usr/bin/cc /usr/bin/c++
C_FLAGS: -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-gnu-alignof-expression -fno-gnu89-i
nline -Wno-cast-function-type
CXX_FLAGS: -fexceptions -funwind-tables -fno-omit-frame-pointer -fno-stack-protector -fno-common -fopenmp -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Wno-gnu-align
of-expression -Wno-cast-function-type
0.49247296061367
0.49198066815734
0.49159581679851
0.4953503748402
0.49523504916579
0.49337223917246
0.49201687611639
0.49092624802142
0.49013123381883
0.48890353925526
$ ./src/tarantool -v && for i in $(seq 1 10); do ./src/tarantool ~/test33.lua; done
Tarantool 2.10.0-beta2-220-gb2c3f868e
Target: Linux-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=ON
Compiler: /usr/bin/cc /usr/bin/c++
C_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type
CXX_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type
0.2586579117924
0.25771971233189
0.26984665263444
0.26222665794194
0.25837187934667
0.26144788879901
0.26106805168092
0.26138027664274
0.26618843432516
0.26360182277858
```

</details>

> Also, I suggest to create ticket with repro and problem description (and mention this tickett here). It issue should be investigated due to possible bug in LuaJIT.

I will, when I dig the root problem. Otherwise, I see no sense in such "dummy" ticket.

> The patch itself LGTM, except 2 nits.

Fixed in the [last version](https://github.com/tarantool/tarantool/commit/b2c3f868e98f5ff35469bf6a7d06a6f25b682f8e).

File: src/lua/digest.lua [hunk 3]
```diff
@@ -149,13 +149,6 @@ setmetatable(PMurHash, {
 
 local CRC32
 local CRC32_methods = {
-    update = function(self, str)
```
Comment (olegrok) [inline src/lua/digest.lua:152]: I guess this line could be removed as well:
https://github.com/tarantool/tarantool/blob/86a2edab6df17a3b0e5ef61ef52a244b33206e51/src/lua/digest.lua#L20

**Review by Buristan — APPROVED** [review]
Hi, thanks for the fixes!
LGTM.

File: src/lua/digest.lua (hunk 3 above)
Comment (igormunkin) [reply src/lua/digest.lua:152]: No, it could not. Please, see its usage in [digest.crc32_update](https://github.com/tarantool/tarantool/blob/b2c3f868e98f5ff35469bf6a7d06a6f25b682f8e/src/lua/digest.lua#L251).
